File: python/factory_design/basic1/plan.py
# author : amit3200 [Amit Singh Sansoya]
from abc import ABC, abstractmethod
class PLAN(ABC):
    rate: float
    block_date: float
    
    def __init__(self):
        pass

    @abstractmethod
    def get_rate(self):
        pass

    def calculate_bill(self,msg,units):
        msg = "{msg} [ Bill ] : {amount}".format(msg = msg,amount = units * self.rate)
        print(msg)
    

File: ruby/factory/basic1/Plan.rb
class Plan
    attr_accessor :rate
    attr_accessor :mode
    def initialize
        @rate = 5.5
        @mode = 'Default'
    end

    def calculate_amount(item, units)
        amount = units * rate
        return amount
    end

    def get_bill(item, units)
        amount = calculate_amount(item, units)
        msg = "#{item} [Bill -> #{@mode}] : #{amount}."
        print_bill(msg)
    end

    def print_bill(msg)
        puts msg 
    end
end